File: medicare-frontend/src/lib/featureFlags.ts
import { useState, useEffect } from 'react';

/**
 * Frontend Feature Flag System.
 * Matches backend defaults and loads environment variables gracefully.
 */
export const DEFAULT_FLAGS = {
  ENABLE_EXPERIMENTAL_SIMULATION: false,
  ENABLE_ADVANCED_ANALYTICS: false,
  ENABLE_NEW_DASHBOARD: false,
  ENABLE_DEBUG_ROUTES: import.meta.env.DEV, // default to true only in local dev
};

export type FeatureFlag = keyof typeof DEFAULT_FLAGS;

class FeatureFlagManager {
  private flags: Record<string, boolean> = { ...DEFAULT_FLAGS };

  constructor() {
    this.loadFromEnv();
  }

  private loadFromEnv() {
    // Vite injects env variables starting with VITE_
    for (const key in this.flags) {
      const envVal = import.meta.env[`VITE_FEATURE_FLAG_${key}`];
      if (envVal !== undefined) {
        this.flags[key] = envVal === 'true' || envVal === '1';
      }
    }
  }

  isEnabled(flag: FeatureFlag): boolean {
    return this.flags[flag] || false;
  }

  getAllFlags(): Record<string, boolean> {
    return { ...this.flags };
  }
}

export const featureFlags = new FeatureFlagManager();

/**
 * React Hook for safely accessing feature flags within components.
 * Useful for conditional rendering of experimental UI features.
 */
export function useFeatureFlag(flag: FeatureFlag): boolean {
  const [enabled, setEnabled] = useState(featureFlags.isEnabled(flag));

  useEffect(() => {
    // Setup listeners if we ever want to dynamically update flags from an API
    setEnabled(featureFlags.isEnabled(flag));
  }, [flag]);

  return enabled;
}

File: medicare-backend/core/feature_flags.py
import os
from typing import Dict, Any

class FeatureFlags:
    """
    Centralized Feature Flag System.
    Safely controls production feature rollout and experimental routing.
    Loads from environment variables dynamically.
    """
    
    def __init__(self):
        # Default fallback states if environment variables are not set
        self._defaults = {
            "ENABLE_EXPERIMENTAL_SIMULATION": False,
            "ENABLE_ADVANCED_ANALYTICS": False,
            "ENABLE_NEW_DASHBOARD": False,
            "ENABLE_DEBUG_ROUTES": False,
        }

    def is_enabled(self, flag_name: str) -> bool:
        """Check if a specific feature flag is enabled."""
        env_val = os.getenv(f"FEATURE_FLAG_{flag_name}")
        if env_val is not None:
            return env_val.lower() in ("true", "1", "yes")
        return self._defaults.get(flag_name, False)

    def get_all_flags(self) -> Dict[str, bool]:
        """Returns the current state of all known feature flags."""
        return {
            flag: self.is_enabled(flag)
            for flag in self._defaults.keys()
        }

feature_flags = FeatureFlags()

def is_feature_enabled(flag_name: str) -> bool:
    """Convenience method to check a feature flag."""
    return feature_flags.is_enabled(flag_name)
